File: SDKs/pixy.py
import requests
from typing import List, Optional, Dict


class PixyImagery:
    def __init__(
        self, base_url: str = "http://localhost:5100", api_key: Optional[str] = None
    ):
        """
        Create a new PixyImagery client

        :param base_url: Base URL of the Pixy Imagery service
        :param api_key: API key for authentication
        """
        self.base_url = base_url
        self.api_key = api_key

    def _construct_url(self, endpoint: str) -> str:
        """
        Construct URL with API key if required

        :param endpoint: Endpoint path
        :return: Full URL with optional API key
        """
        url = f"{self.base_url}{endpoint}"
        if self.api_key:
            url += f"?api_key={self.api_key}"
        return url

    def upload_images(self, images: List) -> Dict[str, List[Dict[str, str]]]:
        """
        Upload one or more images

        :param images: List of images to upload
        :return: Upload result containing image ids and URLs
        """
        url = self._construct_url("/upload-image")
        files = [("images", image) for image in images]

        response = requests.post(url, files=files)

        if response.status_code != 200:
            error_data = response.json()
            raise Exception(error_data.get("error", "Image upload failed"))

        return response.json()

    def get_image_url(
        self,
        image_id: str,
        width: Optional[int] = None,
        height: Optional[int] = None,
        maintain_aspect_ratio: bool = False,
    ) -> str:
        """
        Get image URL by its ID

        :param image_id: ID of the image
        :param width: Desired width
        :param height: Desired height
        :param maintain_aspect_ratio: Whether to maintain aspect ratio
        :return: Image URL
        """
        url = self._construct_url(f"/{image_id}")

        params = {}
        if width:
            params["w"] = width
        if height:
            params["h"] = height
        if maintain_aspect_ratio:
            params["maintain_aspect_ratio"] = "true"

        if params:
            url += "&" + "&".join([f"{key}={value}" for key, value in params.items()])

        return url

    def get_random_image_url(
        self,
        width: Optional[int] = None,
        height: Optional[int] = None,
        maintain_aspect_ratio: bool = False,
    ) -> str:
        """
        Get a random image URL

        :param width: Desired width
        :param height: Desired height
        :param maintain_aspect_ratio: Whether to maintain aspect ratio
        :return: Random image URL
        """
        url = self._construct_url("/random")

        params = {}
        if width:
            params["w"] = width
        if height:
            params["h"] = height
        if maintain_aspect_ratio:
            params["maintain_aspect_ratio"] = "true"

        if params:
            url += "&" + "&".join([f"{key}={value}" for key, value in params.items()])

        return url

File: SDKs/pixy.ts
interface PixyImageryOptions {
  baseUrl?: string;
  apiKey?: string | null;
}

interface ImageUploadResult {
  images: Array<{
    image_id: string;
    url: string;
  }>;
}

interface ResizeOptions {
  width?: number;
  height?: number;
  maintainAspectRatio?: boolean;
}

export class PixyImagery {
  private baseUrl: string;
  private apiKey: string | null;

  /**
   * Create a new PixyImagery client
   * @param options - Configuration options
   */
  constructor(options: PixyImageryOptions = {}) {
    this.baseUrl = options.baseUrl || 'http://localhost:5100';
    this.apiKey = options.apiKey || null;
  }

  /**
   * Construct URL with API key if required
   * @param endpoint - Endpoint path
   * @returns Full URL with optional API key
   */
  private _constructUrl(endpoint: string): string {
    const url = new URL(endpoint, this.baseUrl);
    if (this.apiKey) {
      url.searchParams.set('api_key', this.apiKey);
    }
    return url.toString();
  }

  /**
   * Upload one or more images
   * @param images - Images to upload
   * @returns Upload result
   */
  async uploadImages(images: File[] | FileList): Promise<ImageUploadResult> {
    const formData = new FormData();
    
    // Ensure we're working with an array
    const imageArray = Array.from(images);
    
    // Append each image to FormData
    imageArray.forEach(image => {
      formData.append('images', image);
    });

    const url = this._constructUrl('/upload-image');
    
    const response = await fetch(url, {
      method: 'POST',
      body: formData
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.error || 'Image upload failed');
    }

    return response.json();
  }

  /**
   * Get an image by its ID
   * @param imageId - ID of the image
   * @param options - Resize options
   * @returns Image blob
   */
  async getImage(imageId: string, options: ResizeOptions = {}): Promise<Blob> {
    const url = new URL(this._constructUrl(`/${imageId}`));
    
    if (options.width) url.searchParams.set('w', options.width.toString());
    if (options.height) url.searchParams.set('h', options.height.toString());
    if (options.maintainAspectRatio) url.searchParams.set('maintain_aspect_ratio', 'true');

    const response = await fetch(url.toString());

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.error || 'Failed to retrieve image');
    }

    return response.blob();
  }

  /**
   * Get a random image
   * @param options - Resize options
   * @returns Random image blob
   */
  async getRandomImage(options: ResizeOptions = {}): Promise<Blob> {
    const url = new URL(this._constructUrl('/random'));
    
    if (options.width) url.searchParams.set('w', options.width.toString());
    if (options.height) url.searchParams.set('h', options.height.toString());
    if (options.maintainAspectRatio) url.searchParams.set('maintain_aspect_ratio', 'true');

    const response = await fetch(url.toString());

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.error || 'Failed to retrieve random image');
    }

    return response.blob();
  }
}